File: web/js/game-logic.js
import { TABLE_X, TABLE_Y, TABLE_W, TABLE_H, BALL_RADIUS } from './config.js';

// ── Turn Results ───────────────────────────────────────
export const TURN = {
    CONTINUE: 'continue',
    SWITCH: 'switch',
    FOUL: 'foul',
    WIN: 'win',
    LOSE: 'lose',
};

// ── Game Phases ────────────────────────────────────────
export const PHASE = {
    BREAK: 'break',
    PLAYING: 'playing',
    FINISH: 'finish',
};

// ── Player State ───────────────────────────────────────
export class PlayerState {
    constructor(name) {
        this.name = name;
        this.group = null; // 'solid' | 'stripe'
        this.pocketedBalls = [];
        this.isDone = false;
    }
}

// ── 8-Ball Game Logic ──────────────────────────────────
export class GameLogic {
    constructor(p1Name = 'Player 1', p2Name = 'Player 2') {
        this.players = [new PlayerState(p1Name), new PlayerState(p2Name)];
        this.current = 0;
        this.phase = PHASE.BREAK;
        this.winner = null;
        this.message = `${p1Name}의 브레이크 샷`;
        this.firstHitGroup = null;
    }

    get currentPlayer() { return this.players[this.current]; }
    get otherPlayer() { return this.players[1 - this.current]; }

    static get CUE_START_X() { return TABLE_X + TABLE_W * 0.25; }
    static get CUE_START_Y() { return TABLE_Y + TABLE_H * 0.5; }

    rackPositions() {
        const apexX = TABLE_X + TABLE_W * 0.72;
        const apexY = TABLE_Y + TABLE_H * 0.5;
        const gap = BALL_RADIUS * 2 + 1;
        const rowDx = gap * Math.sqrt(3) / 2;

        const rows = [
            [1],
            [2, 9],
            [3, 8, 10],
            [4, 11, 7, 12],
            [5, 13, 6, 14, 15],
        ];

        const positions = [];
        for (let r = 0; r < rows.length; r++) {
            const row = rows[r];
            const count = row.length;
            for (let c = 0; c < count; c++) {
                const colOff = c - (count - 1) / 2;
                const x = apexX + r * rowDx;
                const y = apexY + colOff * gap;
                positions.push([row[c], x, y]);
            }
        }
        return positions;
    }

    ballInHandPos() {
        return [GameLogic.CUE_START_X, GameLogic.CUE_START_Y];
    }

    onShotEnd(pocketed, firstHitGroup, cuePocketed) {
        // === BREAK ===
        if (this.phase === PHASE.BREAK) {
            if (cuePocketed) {
                this.message = `파울! 큐볼 포켓 — ${this.otherPlayer.name} 차례`;
                this.current = 1 - this.current;
                this.phase = PHASE.PLAYING;
                return TURN.FOUL;
            }
            if (pocketed.length === 0) {
                this.message = `${this.otherPlayer.name}의 차례`;
                this.current = 1 - this.current;
                return TURN.SWITCH;
            }
            // assign groups
            const solids = pocketed.filter(n => n >= 1 && n <= 7).length;
            const stripes = pocketed.filter(n => n >= 9 && n <= 15).length;
            if (solids >= stripes) {
                this.currentPlayer.group = 'solid';
                this.otherPlayer.group = 'stripe';
            } else {
                this.currentPlayer.group = 'stripe';
                this.otherPlayer.group = 'solid';
            }
            this._recordPocketed(pocketed);
            this.phase = PHASE.PLAYING;
            this.message = `그룹 배정! ${this.currentPlayer.name} 계속`;
            return TURN.CONTINUE;
        }

        // === PLAYING ===
        // 8-ball pocketed?
        const eightPocketed = pocketed.includes(8);
        if (eightPocketed) {
            if (this.currentPlayer.isDone) {
                this.phase = PHASE.FINISH;
                this.winner = this.current;
                this.message = `${this.currentPlayer.name} 승리!`;
                return TURN.WIN;
            } else {
                this.phase = PHASE.FINISH;
                this.winner = 1 - this.current;
                this.message = `${this.currentPlayer.name} 패배 (8번 일찍 포켓)`;
                return TURN.LOSE;
            }
        }

        // Foul checks
        if (cuePocketed) {
            this._recordPocketed(pocketed);
            this.message = `파울! 큐볼 포켓 — ${this.otherPlayer.name} 차례`;
            this.current = 1 - this.current;
            return TURN.FOUL;
        }

        // Wrong first hit
        if (firstHitGroup && this.currentPlayer.group) {
            if (firstHitGroup !== this.currentPlayer.group && firstHitGroup !== 'eight') {
                this._recordPocketed(pocketed);
                this.message = `파울! 상대 공 먼저 맞힘 — ${this.otherPlayer.name} 차례`;
                this.current = 1 - this.current;
                return TURN.FOUL;
            }
        }

        this._recordPocketed(pocketed);

        // Check if pocketed own balls
        const ownPocketed = pocketed.filter(n => {
            if (n === 0 || n === 8) return false;
            const g = n <= 7 ? 'solid' : 'stripe';
            return g === this.currentPlayer.group;
        });

        if (ownPocketed.length > 0) {
            this.message = `${this.currentPlayer.name} 계속!`;
            return TURN.CONTINUE;
        }

        this.message = `${this.otherPlayer.name}의 차례`;
        this.current = 1 - this.current;
        return TURN.SWITCH;
    }

    _recordPocketed(nums) {
        for (const n of nums) {
            if (n === 0 || n === 8) continue;
            const g = n <= 7 ? 'solid' : 'stripe';
            for (const p of this.players) {
                if (p.group === g) {
                    if (!p.pocketedBalls.includes(n)) p.pocketedBalls.push(n);
                    p.isDone = p.pocketedBalls.length >= 7;
                }
            }
        }
    }
}

// ── Carom Logic ────────────────────────────────────────
export const CAROM_RESULT = {
    SCORE: 'score',
    MISS: 'miss',
    WIN: 'win',
};

export class CaromLogic {
    constructor(mode = '3cushion', p1 = 'Player 1', p2 = 'Player 2') {
        this.mode = mode;
        this.names = [p1, p2];
        this.scores = [0, 0];
        this.current = 0;
        this.targetScore = mode === '3cushion' ? 10 : 15;
        this.cueBalls = [0, 1]; // P1 uses white(0), P2 uses yellow(1)
    }

    get currentCue() { return this.cueBalls[this.current]; }

    setupBalls() {
        // Returns [number, x, y, colorOverride] for carom balls
        const cx = TABLE_X + TABLE_W / 2;
        const cy = TABLE_Y + TABLE_H / 2;
        const balls = [
            [0, TABLE_X + TABLE_W * 0.25, cy, [248, 248, 248]],      // white
            [1, TABLE_X + TABLE_W * 0.75, cy, [240, 210, 50]],        // yellow
            [2, cx, cy - 50, [210, 30, 30]],                           // red 1
        ];
        if (this.mode === '4ball') {
            balls.push([3, cx, cy + 50, [210, 30, 30]]);              // red 2
        }
        return balls;
    }

    evaluate(cushionHits, ballContacts, cueBallNum) {
        if (this.mode === '3cushion') {
            // must hit 2+ other balls AND 3+ cushions
            const otherContacts = ballContacts.size - (ballContacts.has(cueBallNum) ? 1 : 0);
            if (otherContacts >= 2 && cushionHits >= 3) {
                this.scores[this.current]++;
                if (this.scores[this.current] >= this.targetScore) {
                    return CAROM_RESULT.WIN;
                }
                return CAROM_RESULT.SCORE;
            }
        } else {
            // 4ball: must hit both red balls (2 and 3)
            if (ballContacts.has(2) && ballContacts.has(3)) {
                this.scores[this.current]++;
                if (this.scores[this.current] >= this.targetScore) {
                    return CAROM_RESULT.WIN;
                }
                return CAROM_RESULT.SCORE;
            }
        }
        this.current = 1 - this.current;
        return CAROM_RESULT.MISS;
    }
}

File: client/game/game_logic.py
"""
8볼 게임 규칙 엔진.
상태 머신: BREAK → PLAYING → FINISH
"""
from enum import Enum, auto
from dataclasses import dataclass, field
from typing import Optional
import math
import sys, os
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

from config import TABLE_X, TABLE_Y, TABLE_W, TABLE_H, BALL_RADIUS


class Phase(Enum):
    BREAK   = auto()   # 첫 브레이크 샷 전
    PLAYING = auto()   # 그룹 배정 후 일반 플레이
    FINISH  = auto()   # 게임 종료


class TurnResult(Enum):
    CONTINUE = auto()  # 공 넣음 → 턴 유지
    SWITCH   = auto()  # 못 넣음 → 상대 턴
    FOUL     = auto()  # 파울 → 상대방 볼인핸드
    WIN      = auto()  # 현재 플레이어 승리
    LOSE     = auto()  # 현재 플레이어 패배 (8번 공 실수)


@dataclass
class PlayerState:
    name:   str
    group:  Optional[str] = None   # "solid" | "stripe"
    pocketed_balls: list = field(default_factory=list)

    @property
    def is_done(self) -> bool:
        """자기 그룹 7개를 모두 넣었는지."""
        if self.group is None:
            return False
        if self.group == "solid":
            return len([b for b in self.pocketed_balls if 1 <= b <= 7]) == 7
        else:
            return len([b for b in self.pocketed_balls if 9 <= b <= 15]) == 7


class GameLogic:
    """
    8볼 규칙을 관리한다.
    - phase: BREAK / PLAYING / FINISH
    - current: 현재 턴 플레이어 인덱스 (0 or 1)
    """

    # 큐볼 초기 위치
    CUE_START_X = TABLE_X + TABLE_W * 0.25
    CUE_START_Y = TABLE_Y + TABLE_H * 0.5

    def __init__(self, player1_name: str, player2_name: str):
        self.players  = [PlayerState(player1_name), PlayerState(player2_name)]
        self.current  = 0
        self.phase    = Phase.BREAK
        self.winner: Optional[int] = None   # 0 or 1
        self.message  = ""
        self._first_hit_group: Optional[str] = None

    # ── public ──────────────────────────────────────────

    @property
    def current_player(self) -> PlayerState:
        return self.players[self.current]

    @property
    def other_player(self) -> PlayerState:
        return self.players[1 - self.current]

    def on_shot_end(
        self,
        pocketed: list[int],
        first_hit_group: Optional[str],
        cue_pocketed: bool,
    ) -> TurnResult:
        """
        공이 다 멈춘 후 호출.
        pocketed: 이번 샷에 포켓된 공 번호 목록
        first_hit_group: 큐볼이 처음 맞춘 공의 그룹
        cue_pocketed: 큐볼이 포켓됐는지
        """
        if self.phase == Phase.FINISH:
            return TurnResult.CONTINUE

        # 큐볼 포켓 = 파울
        if cue_pocketed:
            self.message = f"파울: 큐볼이 포켓됐습니다! {self.other_player.name}이 볼인핸드 획득"
            self._switch_turn()
            return TurnResult.FOUL

        real_pocketed = [n for n in pocketed if n != 0]

        # 8번 공 포켓 처리
        if 8 in real_pocketed:
            return self._handle_eight_pocketed(real_pocketed)

        if self.phase == Phase.BREAK:
            return self._handle_break(real_pocketed, first_hit_group)
        else:
            return self._handle_playing(real_pocketed, first_hit_group, cue_pocketed)

    def ball_in_hand_position(self) -> tuple[float, float]:
        """볼인핸드 기본 위치."""
        return self.CUE_START_X, self.CUE_START_Y

    def rack_positions(self) -> list[tuple[int, float, float]]:
        """
        8볼 삼각 랙 배치 위치 반환.
        Returns: [(ball_number, x, y), ...]
        """
        # 랙 꼭짓점: 테이블 오른쪽 3/4 지점
        apex_x = TABLE_X + TABLE_W * 0.72
        apex_y = TABLE_Y + TABLE_H * 0.50
        gap    = BALL_RADIUS * 2 + 1  # 공 지름 + 밀착 여유

        # 표준 8볼 랙 순서 (5행 삼각형)
        order = [
            [1],
            [2, 9],
            [3, 8, 10],
            [4, 11, 7, 12],
            [5, 13, 6, 14, 15],
        ]
        # 행 간격: gap * √3/2  (정삼각형 높이)
        row_dx = gap * math.sqrt(3) / 2
        positions = []
        for row_i, row in enumerate(order):
            row_x = apex_x + row_i * row_dx
            for col_j, ball_num in enumerate(row):
                offset = col_j - (len(row) - 1) / 2
                row_y  = apex_y + offset * gap
                positions.append((ball_num, row_x, row_y))
        return positions

    # ── internal ────────────────────────────────────────

    def _handle_break(self, pocketed: list[int], first_group: Optional[str]) -> TurnResult:
        if not pocketed:
            # 브레이크에서 아무것도 못 넣음 → 턴 교대
            self.message = f"{self.current_player.name}이 턴을 넘깁니다"
            self._switch_turn()
            return TurnResult.SWITCH

        # 브레이크에서 공 넣으면 그룹 배정
        solid_count  = len([n for n in pocketed if 1 <= n <= 7])
        stripe_count = len([n for n in pocketed if 9 <= n <= 15])
        if solid_count >= stripe_count:
            self.current_player.group = "solid"
            self.other_player.group   = "stripe"
        else:
            self.current_player.group = "stripe"
            self.other_player.group   = "solid"
        self.phase = Phase.PLAYING
        self._record_pocketed(pocketed)
        self.message = (
            f"{self.current_player.name} → {self.current_player.group.upper()} | "
            f"{self.other_player.name} → {self.other_player.group.upper()}"
        )
        return TurnResult.CONTINUE

    def _handle_playing(self, pocketed, first_group, cue_pocketed) -> TurnResult:
        my_group    = self.current_player.group
        their_group = self.other_player.group

        # 파울: 첫 번째 맞춘 공이 상대방 공이거나 아무것도 안 맞춤
        if first_group is not None and first_group != my_group and first_group != "eight":
            self.message = f"파울: 상대방 공을 먼저 맞췄습니다"
            self._record_pocketed(pocketed)
            self._switch_turn()
            return TurnResult.FOUL

        # 자기 공 넣음
        my_balls = [n for n in pocketed if self._is_my_ball(n)]
        bad_balls = [n for n in pocketed if self._is_their_ball(n)]

        if bad_balls:
            self.message = f"상대방 공을 넣었습니다! 턴이 교대됩니다"
            self._record_pocketed(pocketed)
            self._switch_turn()
            return TurnResult.FOUL

        if my_balls:
            self._record_pocketed(pocketed)
            if not pocketed:
                # 아무것도 못 넣음
                self.message = f"공을 넣지 못했습니다 → {self.other_player.name} 턴"
                self._switch_turn()
                return TurnResult.SWITCH
            self.message = f"{self.current_player.name}이 {len(my_balls)}개 넣음!"
            return TurnResult.CONTINUE
        else:
            self._record_pocketed(pocketed)
            self.message = f"공을 넣지 못했습니다 → {self.other_player.name} 턴"
            self._switch_turn()
            return TurnResult.SWITCH

    def _handle_eight_pocketed(self, pocketed: list[int]) -> TurnResult:
        if self.phase == Phase.BREAK:
            # 브레이크에서 8번 즉시 넣으면 재랙
            self.message = "브레이크에서 8번 포켓! 다시 랙을 배치합니다"
            return TurnResult.FOUL

        if self.current_player.is_done:
            # 정상 승리
            self.phase  = Phase.FINISH
            self.winner = self.current
            self.message = f"🎱 {self.current_player.name} 승리!"
            return TurnResult.WIN
        else:
            # 아직 자기 공 다 못 넣은 상태에서 8번 포켓 → 패배
            self.phase  = Phase.FINISH
            self.winner = 1 - self.current
            self.message = f"{self.current_player.name}이 8번 공을 너무 일찍 넣었습니다 → {self.other_player.name} 승리!"
            return TurnResult.LOSE

    def _is_my_ball(self, n: int) -> bool:
        g = self.current_player.group
        if g == "solid":
            return 1 <= n <= 7
        return 9 <= n <= 15

    def _is_their_ball(self, n: int) -> bool:
        g = self.other_player.group
        if g is None:
            return False
        if g == "solid":
            return 1 <= n <= 7
        return 9 <= n <= 15

    def _record_pocketed(self, pocketed: list[int]):
        for n in pocketed:
            if n == 0:
                continue
            if self._is_my_ball(n):
                self.current_player.pocketed_balls.append(n)
            elif self._is_their_ball(n):
                self.other_player.pocketed_balls.append(n)

    def _switch_turn(self):
        self.current = 1 - self.current
